### ISSUE TITLE
MacOS killing Hammerspoon due to high CPU usage?

### ISSUE THREAD

**Issue by HerbCSO** [Hammerspoon/hammerspoon#3563]
I have had several occurrences of Hammerspoon getting killed/crashing/not running anymore after some time of letting my system run for a while. I am not really sure what is actually causing this, although I did find the attached diagnostic files that seem to indicate high CPU usage for > 90s, which I am assuming caused MacOS to decide to kill the process (not sure if that's really what it means)?

FWIW there's nothing in `~/Library/Logs/DiagnosticReports`, so I'm really not sure if this is actually a crash or what's happening - all I know is that after some time Hammerspoon is not running anymore (this system runs 24/7). I'd be happy to add some debug logging or whatever might help to figure out the actual root cause here.

This is running on a Mac Mini M2 Pro MacOS Sonoma 14.1.1. I have the same Hammerspoon config running on a 2020 Macbook Pro (Intel i7) running MacOS Ventura 13.5.2 and have *not* run into this issue there.

[Hammerspoon_2023-11-20-181845_Carstens-Mac-mini-2.cpu_resource.diag.txt](https://github.com/Hammerspoon/hammerspoon/files/13488950/Hammerspoon_2023-11-20-181845_Carstens-Mac-mini-2.cpu_resource.diag.txt)
[Hammerspoon_2023-11-27-032947_Carstens-Mac-mini-2.cpu_resource.diag.txt](https://github.com/Hammerspoon/hammerspoon/files/13488951/Hammerspoon_2023-11-27-032947_Carstens-Mac-mini-2.cpu_resource.diag.txt)

**Comment by gab-john** [comment]
Could be related to #3565, have you checked the RAM usage?

**Comment by HerbCSO** [comment]
The problem is that when I notice the issue, it's not hanging, the process is gone, so I can't really check RAM usage anymore at that point.

However, trying the snippet from #3565 in the console does result in high memory utilization and a hang, and also causes a similar diagnostic report to get generated (see
[Hammerspoon_2023-11-28-142229_Carstens-Mac-mini-2.cpu_resource.diag.txt](https://github.com/Hammerspoon/hammerspoon/files/13492344/Hammerspoon_2023-11-28-142229_Carstens-Mac-mini-2.cpu_resource.diag.txt)), so it's possible it's related to that issue. However I do not see Hammerspoon quitting yet, it's just hung now - but it's possible that will take a while longer (I usually notice that Hammerspoon has quit overnight sometime). 

I also don't have a lot of `print` statements in my scripts and spoons, but there are _some_ in there. Maybe that's why it takes several days before it hits this.

**Comment by HerbCSO** [comment]
FWIW I've removed all the `print` statements I could find for now, let's see if it makes a difference.

**Comment by Write** [comment]
I have the same issue on macOS 14.2 RC2, but I have it since Sonoma debut mostly.
Hammerspoon randomly hang at 99% cpu usage and high ram usage. 
Last time I had the bug, it used around 6GB ram.

**Comment by HerbCSO** [comment]
Yeah, I can now confirm that it started with Sonoma. Recently upgraded my Intel MBP to 14.1.2 from 13.5.2 and I started having the same problems there.

**Comment by Grazfather** [comment]
+1 here. Can try to debug next time it occurs.

**Comment by Write** [comment]
Fixed it by removing all logging output to printf and such

> +1 here. Can try to debug next time it occurs.

**Comment by Grazfather** [comment]
> Fixed it by removing all logging output to printf and such
> 
> > +1 here. Can try to debug next time it occurs.

In hammerspoon itself? is the loop in printf itself, or is something still looping?

**Comment by HerbCSO** [comment]
I've now commented out any and all `print` and `printf` statements I could find, and have set all my loggers to `warning` level. 🤞🏼 that works around the issues for now...

**Comment by Write** [comment]
> > Fixed it by removing all logging output to printf and such
> > > +1 here. Can try to debug next time it occurs.
> 
> In hammerspoon itself? is the loop in printf itself, or is something still looping?

In my scripts I removed all printf.
Also I added  this :
```
-- Hide toggling hotkey
hs.hotkey.setLogLevel("warning")
```

And if you still have issue you can try adding that too : 

hs.console.maxOutputHistory(1)
